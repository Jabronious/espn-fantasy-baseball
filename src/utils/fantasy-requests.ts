import { ESPNCookiesDto } from '../models/classes/espn-cookies.dto';
import { env } from '../configuration';
import axios, { AxiosRequestConfig, AxiosResponse } from 'axios';

export class FantasyRequest {
	endpoint: string;
	cookies: ESPNCookiesDto | undefined;
	constructor(leagueId: number, cookies: ESPNCookiesDto | undefined, year: number = new Date().getFullYear()) {
		this.cookies = cookies;
		this.endpoint =
			year === new Date().getFullYear()
				? `${env.ESPN_BASE_ENDPOINT}/${year}/segments/0/leagues/${leagueId}`
				: `${env.ESPN_BASE_ENDPOINT}/leagueHistory/${leagueId}?seasonId=${year}`;
	}

	/*
	 * @Param path: requires that you have "/" at the front of the path
	 * @Param headers: required headers for the request being made
	 * @Param params: optional values to be made with the request
	 */
	async get(path: string = '', headers: unknown = {}, params: unknown = {}): Promise<AxiosResponse> {
		const setCookies = !this.cookies
			? {}
			: { Cookie: `espn_s2=${this.cookies?.espn_s2}; swid=${this.cookies?.swid};` };

		const request: AxiosRequestConfig = {
			url: `${this.endpoint}${path}`,
			method: 'GET',
			params: params,
			headers: { headers, ...setCookies },
		};

		const response = await axios.request(request);
		return response;
	}

	/*
	 * @Param path: requires that you have "/" at the front of the path
	 * @Param headers: required headers for the request being made
	 * @Param params: optional values to be made with the request
	 */
	async post(path: string = '', headers: unknown = {}, params: unknown = {}): Promise<AxiosResponse> {
		const setCookies = !this.cookies
			? {}
			: { Cookie: `espn_s2=${this.cookies?.espn_s2}; swid=${this.cookies?.swid};` };

		const request: AxiosRequestConfig = {
			url: `${this.endpoint}${path}`,
			method: 'POST',
			params: params,
			headers: { headers, ...setCookies },
		};

		const response = await axios.request(request);
		return response;
	}

	setEspnCookies(cookies: ESPNCookiesDto): void {
		this.cookies = cookies;
	}
}
