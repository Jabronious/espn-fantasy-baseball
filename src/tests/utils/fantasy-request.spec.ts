import { FantasyRequest } from '../../utils/fantasy-requests';
import axios from 'axios';

describe('FantasyRequest', () => {
	const fRequest = new FantasyRequest(1234, undefined, 1999);

	it('should be defined', () => {
		expect(fRequest).toBeDefined();
	});

	describe('get', () => {
		it('should accept multiple of the same params', async () => {
			const spy = jest.spyOn(axios, 'request').mockResolvedValueOnce({ data: {} });
			await fRequest.get('', {}, { views: ['1', '2'] });
			expect(spy).toMatchSnapshot();
		});
	});
});
