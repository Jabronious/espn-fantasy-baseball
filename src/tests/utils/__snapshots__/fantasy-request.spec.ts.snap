// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FantasyRequest get should accept multiple of the same params 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "headers": Object {},
        "method": "GET",
        "params": Object {
          "views": Array [
            "1",
            "2",
          ],
        },
        "paramsSerializer": [Function],
        "url": "https://lm-api-reads.fantasy.espn.com/apis/v3/games/flb/seasons/leagueHistory/1234?seasonId=1999",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;
